src/ucp/wireup/wireup_cm.c

```diff
@@ -727,18 +730,39 @@ static unsigned ucp_ep_cm_disconnect_progress(void *arg)
             ucs_assert(ucp_ep->flags & UCP_EP_FLAG_CLOSED);
         }
     } else if (ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED) {
-        ucp_ep_cm_remote_disconnect_progress(ucp_ep);
+        invoke_err_cb  = ucp_ep_cm_remote_disconnect_progress(ucp_ep);
+        ucp_ep->flags &= ~UCP_EP_FLAG_REMOTE_CONNECTED;
+        if (invoke_err_cb) {
+            /* here invoking error callback if it is asked through the flag,
+             * because a user could destroy UCP EP in the error callback */
+            ucp_ep_invoke_err_cb(ucp_ep, UCS_ERR_CONNECTION_RESET);
+        }
+        goto out;
     } else if (ucp_ep->flags & UCP_EP_FLAG_CLOSE_REQ_VALID) {
         /* if the EP is not local connected, the EP has been closed and flushed,
            CM lane is disconnected, complete close request and destroy EP */
         ucs_assert(ucp_ep->flags & UCP_EP_FLAG_CLOSED);
-        close_req = ucp_ep_ext_control(ucp_ep)->close_req.req;
+        ucp_ep->flags &= ~UCP_EP_FLAG_REMOTE_CONNECTED;
+        close_req      = ucp_ep_ext_control(ucp_ep)->close_req.req;
         ucp_ep_local_disconnect_progress(close_req);
+        /* don't touch UCP EP after local disconnect, since it is not valid
+         * anymore */
+        goto out;
+    } else if (ucp_ep->flags & UCP_EP_FLAG_CLOSED) {
+        /* if an EP was closed and not local connected anymore, not failed
+         * and no CLOSE request is set, it means that an EP was disconnected
+         * from a peer */
```

Comment:
does it mean that this function called twice and first time was executed 
```
} else if (ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED) {
```
branch?
Q1 why disconnect event arrived or handled 2 times?
Q2 can we assert here that `UCP_EP_FLAG_ERR_HANDLER_INVOKED` is set? (if peer is disconnected and the ep is not closed yet)

Reply:
> does it mean that this function called twice and first time was executed
> 
> ```
> } else if (ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED) {
> ```
> 
> branch?

no, it means that `ucp_ep_cm_disconnect_cm_lane()` was called from `ucp_ep_close_nbx()`
updated the comment to clarify it

> Q1 why disconnect event arrived or handled 2 times?

no, I don't see that it could be arrived or handled 2 times

> Q2 can we assert here that UCP_EP_FLAG_ERR_HANDLER_INVOKED is set? (if a peer is disconnected and the ep is not closed yet)

we don't expect here that `UCP_EP_FLAG_ERR_HANDLER_INVOKED` is set - added `assert()` for that